lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -262,23 +262,35 @@ def add_credit_card(xml, credit_card, options = {})
           xml.tag! 'NameOnCard', credit_card.first_name
           xml.tag! 'CVNum', credit_card.verification_value if credit_card.verification_value?
 
-          if options[:three_d_secure]
-            three_d_secure = options[:three_d_secure]
-            xml.tag! 'BuyerAuthResult' do
-              xml.tag! 'Status', three_d_secure[:status] unless three_d_secure[:status].blank?
-              xml.tag! 'AuthenticationId', three_d_secure[:authentication_id] unless three_d_secure[:authentication_id].blank?
-              xml.tag! 'PAReq', three_d_secure[:pareq] unless three_d_secure[:pareq].blank?
-              xml.tag! 'ACSUrl', three_d_secure[:acs_url] unless three_d_secure[:acs_url].blank?
-              xml.tag! 'ECI', three_d_secure[:eci] unless three_d_secure[:eci].blank?
-              xml.tag! 'CAVV', three_d_secure[:cavv] unless three_d_secure[:cavv].blank?
-              xml.tag! 'XID', three_d_secure[:xid] unless three_d_secure[:xid].blank?
-            end
-          end
+          add_three_d_secure(options, xml)
 
           xml.tag! 'ExtData', 'Name' => 'LASTNAME', 'Value' =>  credit_card.last_name
         end
       end
 
+      def add_three_d_secure(options, xml)
+        if options[:three_d_secure]
+          three_d_secure = options[:three_d_secure]
+          xml.tag! 'BuyerAuthResult' do
+            xml.tag! 'Status', three_d_secure[:status] unless three_d_secure[:status].blank?
+            xml.tag! 'AuthenticationId', three_d_secure[:authentication_id] unless three_d_secure[:authentication_id].blank?
+            xml.tag! 'AUTHENTICATION_STATUS', authentication_status(three_d_secure)
```

Comment:
do we need to check if this is `blank?` too?

Reply:
The doc you linked above is using a different key format, for instance `AuthenticationId` is written as `AUTHENTICATION_ID` in your doc. This mean that `AUTHENTICATION_STATUS` probably needs to be changed to `AuthenticationStatus`

Reply:
i was doing a check like this before:
```
return three_d_secure[:authentication_response_status] unless three_d_secure[:authentication_response_status].blank?
        three_d_secure[:directory_response_status] unless three_d_secure[:directory_response_status].blank?
```
but Filipe commented that if neither of them are present, it's going to return `nil`, so we can do a check this way:
```three_d_secure[:authentication_response_status] || three_d_secure[:directory_response_status]```

do you think a `blank?` check is still necessary?

Reply:
Regarding the `AUTHENTICATION_STATUS`:
I agree that remote tests should be done. On master there are 10 failing tests when you run the payflow remote tests, so I havent written any new ones
I followed the capital convention like `ECI`, `CAVV` and `XID` because we send them and we dont send `PAReq`, `ACSUrl` and `AuthenticationId`

but how should i go about the remote tests?

Reply:
Why do we have two keys for the same status? 

It seems to me like we should not be sending `xml.tag! 'AUTHENTICATION_STATUS'` if it contains a blank status (to be consistent with the other keys).

Reply:
> I followed the capital convention like ECI, CAVV and XID because we send them

The sample request in your linked docs is not sending XML. Is there another doc with an xml example?
```
VENDOR=MerchantUserID&PARTNER=PayPal&USER=UserIDIfAvailOrSameAsVendor&PWD=Pwd4Payflow&TENDER=C&TRXTYPE=S&TENDER=C&ACCT=5555555555554444&EXPDATE=0308&AMT=123.00&AUTHENTICATION_ID[20]=8d4d5ed66ac6e6faac6d&CAVV[28]=OTJlMzViODhiOTllMjBhYmVkMGU=&AUTHENTICATION_STATUS[1]=1&ECI[1]=5&XID[28]=YjM0YTkwNGFkZTI5YmZmZWE1ZmY&THREEDSVERSION[5]=1.0.2
```

Reply:
This document has only `AUTHENTICATION_STATUS` in TABLE 3.4 Transaction response parameters... but nothing in the request.  https://www.paypalobjects.com/webstatic/en_US/developer/docs/pdf/pp_payflowpro_xmlpay_guide.pdf

Reply:
`authentication_response_status` is the MPI response from cmpi_authenticate response whereas `directory_response_status` is from cmpi_lookup, and if the authentication was frictionless we'll have `directory_response_status`, but if it was with a challenge then we'll have `authentication_response_status` as the final response that PayPal wants
>Value returned by MPI indicating if authentication was successful, attempted or failed.

That's why I believe we need both

and I see your point about the empty tag, i'll bring the `blank?` checks back

Reply:
Ryan from paypal sent this:
```
<Tender>
              <Card>
                <CardType>Visa</CardType>
                <CardNum>[FILTERED]</CardNum>
                <ExpDate>202001</ExpDate>
                <NameOnCard>CHRISTOPHER</NameOnCard>
                <CVNum>[FILTERED]</CVNum>
                <BuyerAuthResult>
                  <ECI>05</ECI>
                  <CAVV>AAABBFiCBAAAAFhBkoIEAAAAAAA=</CAVV>
                  <XID>QUlxdno2WEduZE5qNlU0RU1NZTA=</XID>
                </BuyerAuthResult>
                <ExtData Name="LASTNAME" Value="D KING"/>
                <ExtData Name="THREEDSVERSION" Value=""/>
                <ExtData Name="DSTRANSACTIONID" Value=""/>
              </Card>
            </Tender>
```
but after that he said:
>you’re not passing us `AUTHENTICATION_STATUS`
and linked the documentation in the PR description

Reply:
```xml
<element name = "BuyerAuthResult">
  <complexType content = "elementOnly">
    <sequence>
      <element name = "Status" type = "BuyerAuthStatusEnum"/>
      <element name = "AuthenticationId" type = "Base64Sha1StringType" minOccurs="0"
maxOccurs="1"/>
      <element name = "PAReq" type = "BuyerAuthMesgType" minOccurs="0" maxOccurs="1" />
      <element name = "ACSUrl" type = "uriReference" minOccurs="0" maxOccurs="1" />
      <element name = "ECI" type = "ECIType" minOccurs="0" maxOccurs="1" />
      <element name = "CAVV" type = "Base64Sha1StringType" minOccurs="0" maxOccurs="1" />
      <element name = "XID" type = "Base64Sha1StringType" minOccurs="0" maxOccurs="1" />
    </sequence>
  </complexType>
</element>
```
This is what we are supposed to send, so `AUTHENTICATION_STATUS` is just `Status`

Reply:
that's exactly what we're sending already, I think that xml doc is out of date, it's from 2013

Reply:
so I ran the remote tests with `AUTHENTICATION_STATUS` and `AuthenticationStatus`, they both pass